brushfire-core/src/main/scala/com/stripe/brushfire/Injections.scala

```diff
@@ -135,58 +157,70 @@ object JsonInjections {
                 obj.put("children", toJsonNode(child)(nodeJsonNodeInjection))
                 ary.add(obj)
             }
-            ary
+
+            val obj = JsonNodeFactory.instance.objectNode
+            obj.put("annotation", toJsonNode(annotation))
+            obj.put("splits", ary)
+            obj
         }
 
-        def tryChild(node: JsonNode, property: String) = Try {
-          val child = node.get(property)
-          assert(child != null, property + " != null")
-          child
+        private def invertLeaf(n: JsonNode, indexNode: JsonNode): Try[LeafNode[K, V, T, A]] = {
+          for {
+            index <- fromJsonNode[Int](indexNode)
+            target <- fromJsonNode[T](n.get("distribution"))
+            annotation <- fromJsonNode[A](n.get("annotation"))
+          } yield {
+            LeafNode(index, target, annotation)
+          }
         }
 
-        override def invert(n: JsonNode) = {
-          Option(n.get("leaf")) match {
-            case Some(indexNode) => fromJsonNode[Int](indexNode).flatMap { index =>
-              fromJsonNode[T](n.get("distribution")).map { target =>
-                LeafNode(index, target)
+        private def invertSplit(n: JsonNode): Try[SplitNode[K, V, T, A]] = {
+          import Monad.scalaTry // needed for Applicative.sequence
+          for {
+            annotation <- fromJsonNode[A](n.get("annotation"))
+            splitsNode <- tryChild(n, "splits")
+            splits <- Applicative.sequence {
```

Comment:
Seems pretty reasonable to me. I've never looked at algebird's Monad stuff before, thanks for giving me an excuse to :)